### PR TITLE
test: use higher value and per-platform assert in cnetaddr link-local test

### LINKED ISSUE #21682: intermittent unit test failure: test/net_tests.cpp(311): error: in "net_tests/cnetaddr_basic": check addr_str == scoped_addr || addr_str == "fe80:0:0:0:0:0:0:1" has failed
https://cirrus-ci.com/task/5470546704990208?logs=ci#L4527

### PR DESCRIPTION
**Pull request by jonatack**
This patch improves the cnetaddr link-local test in `net_tests` by making it platform-specific and by using a higher scoped id value to alleviate recent CI issues.

Closes #21682.

### CODE REVIEW

**Comment by practicalswift** [comment]
The intermittently failing `BOOST_CHECK` is:

https://github.com/bitcoin/bitcoin/blob/9712f75746e3da73471da2e23a4bfc1382c69308/src/test/net_tests.cpp#L311-L312

I don't understand why we test if `ToString()` output includes `%zone_index`: it clearly doesn't on some platforms, so we cannot rely on it anyways. Then why test it? :)

**Comment by jonatack** [comment]
My understanding when working on this was that Apple might be going in its own direction with this starting with macOS 10.14 but it still seemed worthwhile to test our code on platforms like Linux and Windows where we can.

**Comment by jonatack** [comment]
(Without the fallback, only the macOS 10.14 CI was failing. So coverage on the other CIs seemed good to have.)

**Comment by practicalswift** [comment]
> My understanding when working on this was that Apple might be going in its own direction with this starting with macOS 10.14 but it still seemed worthwhile to test our code on platforms like Linux and Windows where we can.

But the test is not conditional on platform, so AFAICT we're testing that `ToString` either contains `%zone_id` (the `addr_str == scoped_addr` case) or that it _doesn't_ contain `%zone_id` (the `addr_str == "fe80:0:0:0:0:0:0:1"` case)? :)

https://github.com/bitcoin/bitcoin/blob/9712f75746e3da73471da2e23a4bfc1382c69308/src/test/net_tests.cpp#L310-L312

In other words: we're effectively not testing anything meaningful in a test that is intermittently failing, no? :)

If so, why not simply remove it? :)

**Comment by jonatack** [comment]
If you test only the fallback, I found that the change fails on all CIs except the macOS 10.14 one:
```diff
-    BOOST_CHECK(addr_str == scoped_addr || addr_str == "fe80:0:0:0:0:0:0:1");
+    BOOST_CHECK(addr_str == "fe80:0:0:0:0:0:0:1");
```

$ src/test/test_bitcoin -t net_tests
Running 14 test cases...
test/net_tests.cpp(311): error: in "net_tests/cnetaddr_basic": check addr_str == "fe80:0:0:0:0:0:0:1" has failed

*** 1 failure is detected in the test module "Bitcoin Core Test Suite"
```

**Comment by MarcoFalke** [comment]
Does this actually fix the issue or make it only less likely?

**Comment by jonatack** [comment]
As far as I know, the issue began appearing in the CI only this week, so it may be from some recent change in the CI setup, as the test value wasn't very high. The CI is green here, so re-pushing now with a very high value. If that causes a failure down the line, I guess we could drop the coverage as suggested above.

**Comment by practicalswift** [comment]
@jonatack

I didn't get if you agree with my statement that the intermittently failing `BOOST_CHECK` is basically doing:

```
BOOST_CHECK(to_string_does_contain_zone_id || to_string_does_not_contain_zone_id);
```

And if so, what the point of keeping it around? :)

---

FWIW, patch against current `master` to show the issue more clearly:

```diff
diff --git a/src/test/net_tests.cpp b/src/test/net_tests.cpp
index 8eab26f3d..e7c3475ef 100644
--- a/src/test/net_tests.cpp
+++ b/src/test/net_tests.cpp
@@ -308,8 +308,9 @@ BOOST_AUTO_TEST_CASE(cnetaddr_basic)
     BOOST_REQUIRE(addr.IsIPv6());
     BOOST_CHECK(!addr.IsBindAny());
     const std::string addr_str{addr.ToString()};
-    BOOST_CHECK(addr_str == scoped_addr || addr_str == "fe80:0:0:0:0:0:0:1");
-    // The fallback case "fe80:0:0:0:0:0:0:1" is needed for macOS 10.14/10.15 and (probably) later.
+    const bool to_string_does_contain_zone_id = addr_str == scoped_addr;
+    const bool to_string_does_not_contain_zone_id = addr_str == "fe80:0:0:0:0:0:0:1";
+    BOOST_CHECK(to_string_does_contain_zone_id || to_string_does_not_contain_zone_id);
     // Test that the delimiter "%" and default zone id of 0 can be omitted for the default scope.
     BOOST_REQUIRE(LookupHost(link_local + "%0", addr, false));
     BOOST_REQUIRE(addr.IsValid());
```

**Comment by jonatack** [comment]
@practicalswift It looks like I should document the test better to communicate the context and information. I'll propose that separately.

**Comment by fanquake** [comment]
> It looks like I should document the test better to communicate the context and information. I'll propose that separately.

Why can't that be done as part of this change?

**Comment by jonatack** [comment]
> > It looks like I should document the test better to communicate the context and information. I'll propose that separately.
> 
> Why can't that be done as part of this change?

This change was just to update a test that was stable until just recently. But sure, can propose it here it that doesn't hold things up.

**Comment by jonatack** [comment]
Ok. The CI has been green, and #21696 showed that macOS remains the only CI that fails the `scoped_addr` case.

Let's try (a) improving the documentation and (b) tightening the `scoped_addr` assertion to be XOR instead of OR on the two cases.

**Comment by jonatack** [comment]
Great--the per-platform scoped address test is green along with the previous two CI runs.  Pushing an updated version with `BOOST_CHECK( == )` replaced by `BOOST_CHECK_EQUAL( , )` for a 4th CI run. I hope this addresses the feedback above while maintaining coverage and adding clarity and context.

**Comment by MarcoFalke** [comment]
Needs rebase after https://github.com/bitcoin/bitcoin/pull/21689#issuecomment-821794629

**Comment by jonatack** [comment]
I should have updated the test in a separate PR from the one fixing the reported new issue.

**Comment by jonatack** [comment]
> If so, why not simply remove it? :)

I disagree with the approach of dropping existing test coverage when it is not understood.  Better to improve the test and documentation in that case and I spent time re-verifying and working to do that.  Unfortunately, it's an order of magnitude more work to refute statements like "not very meaninfgul" in the #21689 PR title and here, than it is to write them and just erase coverage and (I think) valuable context.

**Comment by jonatack** [comment]
Rebased, added 6b0e879 for context since this change is not adding a test but hopefully improving it.

File: src/test/net_tests.cpp [hunk 1]
```diff
@@ -300,16 +300,21 @@ BOOST_AUTO_TEST_CASE(cnetaddr_basic)
 
     // IPv6, scoped/link-local. See https://tools.ietf.org/html/rfc4007
     // We support non-negative decimal integers (uint32_t) as zone id indices.
-    // Test with a fairly-high value, e.g. 32, to avoid locally reserved ids.
+    // Test with a high value, e.g. 4096, to avoid any local ids in use.
     const std::string link_local{"fe80::1"};
-    const std::string scoped_addr{link_local + "%32"};
+    const std::string scoped_addr{link_local + "%4096"};
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/test/net_tests.cpp:305]: Good question. From what I understand, the scope ids / zone indices depend on the number of local network interfaces when there are multiple on a single link-local address. Whereas many machines may have the first 8 or 16 decimal integer scope ids in use, the previous test value of 32 seems to be on the high end of the range. I'm not aware of one that is guaranteed to be free, but a value in the thousands (or tens of thousands) might be reasonably worth trying.

Resources:
- https://en.wikipedia.org/wiki/IPv6_address#zone_index
- https://tools.ietf.org/html/rfc4007#section-6
- https://tools.ietf.org/html/rfc4007#section-11.2

**Comment by jonatack** [comment]
@practicalswift does this now address your feedback?

**Comment by practicalswift** [comment]
> @practicalswift does this now address your feedback?

The `BOOST_CHECK(to_string_does_contain_zone_id || to_string_does_not_contain_zone_id)` issue has been addressed.

I think the issue @MarcoFalke raised about the test outcome being dependent on the user's local configuration remains unfortunately.

File: src/test/net_tests.cpp [hunk 2]
```diff
@@ -300,13 +300,21 @@ BOOST_AUTO_TEST_CASE(cnetaddr_basic)
...
     BOOST_REQUIRE(addr.IsIPv6());
     BOOST_CHECK(!addr.IsBindAny());
+    const std::string addr_str{addr.ToString()};
+#ifdef MAC_OSX
+    // Expect macOS to return the address in this format, without zone ids.
+    BOOST_CHECK_EQUAL(addr_str, "fe80:0:0:0:0:0:0:1");
+#else
+    // Expect normal scoped address functionality on the other platforms.
+    BOOST_CHECK_EQUAL(addr_str, scoped_addr);
+#endif
```
Comment (MarcoFalke) [inline src/test/net_tests.cpp:317]:
```suggestion
    BOOST_CHECK(addr.ToString().starts_with("fe80:0:0:0:0:0:0:1"));
```

Maybe just check the prefix if the suffix is unstable?

**Comment by jonatack** [comment]
This test and informational context involved some trial and error and time to obtain, and I think it's a shame to throw out the coverage and information, particularly as it seems to be not well-known info and I tried to provide context about that. However, it seems like I'm swimming upstream to provide this coverage and information.

**Comment by jonatack** [comment]
The scoped id/zone indices functionality appears to have been dropped entirely by #21756.

**Comment by jonatack** [comment]
The `Up for grabs` label on this pull can be removed following the merge of #21985.

**Pull request closed without merging**